web_ui/packages/smart-tools/src/utils/utils.ts

```diff
@@ -36,3 +36,71 @@ export const approximateShape = (CV: OpenCVTypes.cv, contour: OpenCVTypes.Mat, i
 
     return newContour;
 };
+
+export const concatFloat32Arrays = (arrays: Float32Array[]) => {
+    const totalLength = arrays.reduce((c, a) => c + a.length, 0);
+    const result = new Float32Array(totalLength);
+
+    arrays.reduce((offset, array) => {
+        result.set(array, offset);
+        return offset + array.length;
+    }, 0);
+
+    return result;
+};
+
+export const loadSource = async (source: string, cacheKey = 'general'): Promise<Response | undefined> => {
+    if (!caches) {
+        return await self.fetch(source);
+    }
+
+    const cache = await caches.open(cacheKey);
+
+    if (!(await cache.match(source))) {
+        await cache.put(source, await self.fetch(source));
+    }
+
+    return cache.match(source);
+};
+
+export const stackPlanes = (CV: OpenCVTypes.cv, mat: OpenCVTypes.Mat) => {
+    let stackedPlanes: OpenCVTypes.Mat[] = [];
+    let matPlanes: OpenCVTypes.MatVector | null = null;
+
+    try {
+        matPlanes = new CV.MatVector();
+        CV.split(mat, matPlanes);
+
+        stackedPlanes = Array.from(Array(mat.channels()).keys()).map((index) => {
+            // This won't happen, but matPlanes is mutable for the finally block.
+            if (!matPlanes) {
+                throw 'Lost track of matPlanes through loop';
+            }
+
+            return matPlanes.get(index);
+        });
+
+        return concatFloat32Arrays(stackedPlanes.map((m) => m.data32F));
+    } finally {
+        stackedPlanes.map((p) => p.delete());
+        matPlanes?.delete();
+    }
+};
+
+export const isPolygonValid = (polygon: Polygon | null): boolean => {
+    const MINIMUM_POLYGON_AREA = 4;
+
+    if (!polygon) return false;
+
+    const points = polygon.points;
+    if (points.length < 3) return false;
+
+    let area = 0;
+    for (let i = 0, j = points.length - 1; i < points.length; j = i++) {
+        area += (points[j].x + points[i].x) * (points[j].y - points[i].y);
+    }
+
+    area = Math.abs(area / 2);
```

Comment:
This looks wrong to me: we aren't subtracting overlapping areas here, so you could draw a very thin angled rectangle that would have a small area but this function would return a very large area.
I'd prefer if we keep the old implementation.